### ISSUE TITLE
normalisation data for --passthrough

### ISSUE THREAD

**Issue by DeLub** [librespot-org/librespot#655]
Hi,

Currently the normalisation options seem to have no effect when --passthrough is used, which makes sense. Is there an option when piping into ffmpeg with --passthrough to also receive the replaygain values? Can they be inserted at the start of the track in the headers?

**Comment by philippe44** [comment]
It would require a more extensive rewrite of the headers which purposely very minimal at that moment. I can look into that later once the updated passthrough option has been adopted and feedback is good.

**Comment by roderickvd** [comment]
@philippe44 I took the liberty of assigning this one to you, if you'd have the time to spare and think this is a good idea.

**Comment by roderickvd** [comment]
@philippe44 is this something you would want to look into?

**Comment by philippe44** [comment]
> @philippe44 is this something you would want to look into?
>
Do you think this is an important feature? 

My last attempt to change something in librespot was rejected (https://github.com/librespot-org/librespot/discussions/741 and https://github.com/philippe44/librespot/tree/HTTP-sink) and I can understand why but as I'm not good (to say the least) at Rust, these modification are a lot of hard work for me and I lost a bit my motivation since last time. 

I will maintain when needed my original contribution, but I'm not sure I'm willing to take the challenge for this replaygain option, especially because it's potentially a fair bit of extra work as one now need to dig a level lower into headers. Give me a couple of weeks to think about that.

**Comment by roderickvd** [comment]
> > @philippe44 is this something you would want to look into?
> 
> Do you think this is an important feature?

I feel the best way I can answer is by principle: I think that the passthrough decoder should do exactly what's on the tin, namely, passing through the audio files as they are.

Taking that as a guiding principle, I think it would be good that if there's normalisation data in the incoming audio file, that should also be in the passed through audio files.

TL;DR yes I think so for feature-completeness.
 
> My last attempt to change something in librespot was rejected (#741 and https://github.com/philippe44/librespot/tree/HTTP-sink) and I can understand why but as I'm not good (to say the least) at Rust, these modifications are a lot of hard work for me and I lost a bit my motivation since last time.

I am sorry that you feel that way. I never intended for that.

> I will maintain, when needed, my original contribution, but I'm not sure I'm willing to take the challenge for this replaygain option, especially because it's potentially a fair bit of extra work as one now need to dig a level lower into headers (and in Rust...). Give me a couple of weeks to think about that.

Thanks for that commitment of maintainership.

If it's of any help, these lines show where the normalisation data is at in Ogg Vorbis:

https://github.com/librespot-org/librespot/blob/68bec41e08564187e27d8c132b9b7b8274579819/playback/src/player.rs#L217-L234

**Comment by philippe44** [comment]
Thanks for the answer - I understand the statement "the passthrough decoder should do exactly what's on the tin, namely, passing through the audio files as they are" and that's what I'm doing but I think the OP wants something different. 

AFAII, the original problem in librespot was that replaygain was ignored by the decoder and a PR was made in 2018 to apply gain on PCM samples. Then an option was added to add a manual replaygain tweak (and maybe a few other options I don't know). So my understanding is that the ask is to use that "extra" replaygain and to *add/insert* it in re-written headers

**Comment by DeLub** [comment]
My suggested feature request was to have the original replaygain values that go along with the original, untouched audio data.

**Comment by philippe44** [comment]
So I'm a bit confused then b/c the replaygain, as set in the ogg headers given by spotify is passthrough as well. I don't remove/touch it

**Comment by philippe44** [comment]
@DeLub: can you comment? Do you see something different?

**Comment by DeLub** [comment]
I'm sorry for the late reply.

When I created this request, I was not able to normalise with ffmpeg. That's why I put forward this request.

However, at this moment I'm not using Spotify anymore. I switched to Apple Music as the Apple One subscription saves me money, and actually AM and Spotify are quite comparable in my opinion.

**Comment by roderickvd** [comment]
OK, let's close this then. Thanks for getting back.

**Comment by roderickvd** [comment]
To revisit this for posterity:

Having worked on the audio data handling, I now understand where this is coming from -- and @philippe44 was right. Spotify doesn't store the ReplayGain values in the Vorbis comments where you'd expect them (and which are passed through 1-on-1), but in a custom Ogg packet before it.

If you want to have the ReplayGain fields as Vorbis comments, then you'd have to modify the stream to insert them. This is currently not what the passthrough decoder does nor should, probably. Modifying is not passing through.
